src/language-services/validation.ts

```diff
@@ -127,8 +180,151 @@ function maybeParseErrorToDiagnostic(error: PQP.ParseError.ParseError): Diagnost
+    };
+
+    return PQP.Traverse.tryTraverseXor(
+        traversalState,
+        nodeIdMapCollection,
+        root,
+        PQP.Traverse.VisitNodeStrategy.BreadthFirst,
+        visitNode,
+        PQP.Traverse.expectExpandAllXorChildren,
+        undefined,
+    );
+}
+
+function keyValuePairsToSymbols(
+    keyValuePairs: ReadonlyArray<
+        PQP.NodeIdMapIterator.KeyValuePair<PQP.Language.Ast.GeneralizedIdentifier | PQP.Language.Ast.Identifier>
+    >,
+): DocumentSymbol[] {
+    const symbols: DocumentSymbol[] = [];
+    keyValuePairs.forEach(value => {
+        const range: Range = LanguageServiceUtils.tokenRangeToRange(value.key.tokenRange);
+        symbols.push({
+            kind: SymbolKind.Variable,
+            name: value.keyLiteral,
+            range,
+            selectionRange: range,
+        });
+    });
+
+    return symbols;
+}
+
+function visitNode(state: TraversalState, currentXorNode: PQP.TXorNode): void {
+    let symbols: DocumentSymbol[] | undefined = undefined;
+
+    // TODO: Validate that "in" variable exists
+    switch (currentXorNode.node.kind) {
+        case PQP.Language.Ast.NodeKind.LetExpression:
+            const letMembers: ReadonlyArray<PQP.NodeIdMapIterator.KeyValuePair<
+                PQP.Language.Ast.Identifier
+            >> = PQP.NodeIdMapIterator.letKeyValuePairs(state.nodeIdMapCollection, currentXorNode);
+            if (letMembers.length > 1) {
+                symbols = keyValuePairsToSymbols(letMembers);
+            }
+            break;
+
+        case PQP.Language.Ast.NodeKind.RecordExpression:
+        case PQP.Language.Ast.NodeKind.RecordLiteral:
+            const recordFields: ReadonlyArray<PQP.NodeIdMapIterator.KeyValuePair<
+                PQP.Language.Ast.GeneralizedIdentifier
+            >> = PQP.NodeIdMapIterator.recordKeyValuePairs(state.nodeIdMapCollection, currentXorNode);
+            if (recordFields.length > 1) {
+                symbols = keyValuePairsToSymbols(recordFields);
+            }
+            break;
+
+        case PQP.Language.Ast.NodeKind.Section:
+            const sectionMembers: ReadonlyArray<PQP.NodeIdMapIterator.KeyValuePair<
+                PQP.Language.Ast.Identifier
+            >> = PQP.NodeIdMapIterator.sectionMemberKeyValuePairs(state.nodeIdMapCollection, currentXorNode);
+            if (sectionMembers.length > 1) {
+                symbols = keyValuePairsToSymbols(sectionMembers);
+            }
+            break;
+
+        default:
+    }
+
+    if (symbols) {
+        // Look for duplicate member/variable/field names
+        state.result.push(...identifyDuplicateSymbols(state, symbols));
+    }
+}
+
+function identifyDuplicateSymbols(state: TraversalState, symbols: DocumentSymbol[]): Diagnostic[] {
+    const result: Diagnostic[] = [];
+    const seenSymbols: Map<string, DocumentSymbol[]> = new Map<string, DocumentSymbol[]>();
+
+    symbols.forEach(symbol => {
+        if (seenSymbols.has(symbol.name)) {
```

Comment:
I do something like the below. It saves a double key lookup, but I can't say if that even effects performance since JavaScript can be wonky that way.

```
const symbolsForName: DocumentSymbol[] | undefined = seenSymbols.get(...)
if (symbolsForName) { symbolsForName.push(...) } else { seenSymbols.set(...) }
```